pupil_src/shared_modules/annotations.py

```diff
@@ -10,220 +10,297 @@
-        if self.buttons:
-            for b in self.buttons:
-                self.g_pool.quickbar.remove(b)
-            self.buttons = []
-
-    def append_annotation(self):
-        self.annotation_definitions.append(
-            (self.new_annotation_name, self.new_annotation_hotkey)
+
+    def _clear_button_quickbars(self):
+        # only call this from deinit_ui()
+        for buttons in self._definition_to_buttons.values():
+            self._remove_button_quickbar(buttons.quickbar)
+
+    @abstractmethod
+    def fire_annotation(self, annotation_definition):
+        pass
+
+    def _on_add_annotation_clicked(self):
+        # new_annotation_label and hotkey are set by the ui input fields
+        self._add_annotation_definition(
+            self._new_annotation_label, self._new_annotation_hotkey
         )
-        self.update_buttons()
 
-    def remove_annotation(self, annotation):
-        try:
-            self.annotation_definitions.remove(annotation)
-        except ValueError:
-            print(annotation, self.annotation_definitions)
-        self.update_buttons()
+    def _add_annotation_definition(self, annotation_label, hotkey):
+        annotation_definition = AnnotationDefinition(
+            label=annotation_label, hotkey=hotkey
+        )
+        if annotation_definition in self._definition_to_buttons:
+            return
+        button_quickbar = self._create_button_quickbar(annotation_definition)
+        button_menu = self._create_button_menu(annotation_definition)
+        annotation_ui = self._AnnotationButtons(button_quickbar, button_menu)
+        self._definition_to_buttons[annotation_definition] = annotation_ui
+
+        self._append_button_quickbar(button_quickbar)
+        self._append_button_menu(button_menu)
+
+    def _create_button_quickbar(self, annotation_definition):
+        def make_fire(_):
+            self.fire_annotation(annotation_definition.label)
+
+        return ui.Thumb(
+            annotation_definition.label,
+            setter=make_fire,
+            getter=lambda: False,
+            label=annotation_definition.hotkey,
+            hotkey=annotation_definition.hotkey,
+        )
 
-    def fire_annotation(self, annotation_label):
-        t = self.g_pool.get_timestamp()
-        logger.info('"{}"@{}'.format(annotation_label, t))
-        # you may add more field to this dictionary if you want.
-        notification = {
-            "subject": "annotation",
-            "label": annotation_label,
-            "timestamp": t,
-            "duration": 0.0,
-            "record": True,
-        }
-        self.notify_all(notification)
+    def _create_button_menu(self, annotation_definition):
+        def make_remove():
+            self._remove_annotation_info(annotation_definition)
+
+        label = annotation_definition.label
+        hotkey = annotation_definition.hotkey
+        return ui.Button(label=label + " <" + hotkey + ">", function=make_remove)
```

Comment:
```suggestion
        return ui.Button(label="{} <{}>".format(label, hotkey), function=make_remove)
```
We usually prefer string-formatting over manual formatting.

I am looking forward to not having to support Python 3.5 anymore. Then we can do this:
```python
f"{label} <{hotkey}>"
```